// passport-local-mongoose/dist/lib/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TooManyAttemptsError = exports.AttemptTooSoonError = exports.NoSaltValueStoredError = exports.UserExistsError = exports.MissingPasswordError = exports.MissingUsernameError = exports.IncorrectPasswordError = exports.IncorrectUsernameError = exports.AuthenticationError = void 0;
class AuthenticationError extends Error {
    constructor(message) {
        super(message);
        this.name = 'AuthenticationError';
        Error.captureStackTrace(this, this.constructor);
    }
}
exports.AuthenticationError = AuthenticationError;
class IncorrectUsernameError extends AuthenticationError {
    constructor(message) {
        super(message);
        this.name = 'IncorrectUsernameError';
    }
}
exports.IncorrectUsernameError = IncorrectUsernameError;
class IncorrectPasswordError extends AuthenticationError {
    constructor(message) {
        super(message);
        this.name = 'IncorrectPasswordError';
    }
}
exports.IncorrectPasswordError = IncorrectPasswordError;
class MissingUsernameError extends AuthenticationError {
    constructor(message) {
        super(message);
        this.name = 'MissingUsernameError';
    }
}
exports.MissingUsernameError = MissingUsernameError;
class MissingPasswordError extends AuthenticationError {
    constructor(message) {
        super(message);
        this.name = 'MissingPasswordError';
    }
}
exports.MissingPasswordError = MissingPasswordError;
class UserExistsError extends AuthenticationError {
    constructor(message) {
        super(message);
        this.name = 'UserExistsError';
    }
}
exports.UserExistsError = UserExistsError;
class NoSaltValueStoredError extends AuthenticationError {
    constructor(message) {
        super(message);
        this.name = 'NoSaltValueStoredError';
    }
}
exports.NoSaltValueStoredError = NoSaltValueStoredError;
class AttemptTooSoonError extends AuthenticationError {
    constructor(message) {
        super(message);
        this.name = 'AttemptTooSoonError';
    }
}
exports.AttemptTooSoonError = AttemptTooSoonError;
class TooManyAttemptsError extends AuthenticationError {
    constructor(message) {
        super(message);
        this.name = 'TooManyAttemptsError';
    }
}
exports.TooManyAttemptsError = TooManyAttemptsError;
